dfaker.py

```diff
@@ -0,0 +1,389 @@
+	return offset
+
+def make_gaps(params, time_gluc):
+	gaps = random.randint(1 * params['num_days'], 3 * params['num_days']) # amount of gaps	
+	for _ in range(gaps):
+		gap_length = random.randint(6, 36) # length of gaps in 5-min segments
+		start_index = random.randint(0, len(time_gluc))	
+		if start_index + gap_length > len(time_gluc):
+			end_index = len(time_gluc) - 5
+		else:
+			end_index = start_index + gap_length
+		time_gluc = np.delete(time_gluc, time_gluc[start_index:end_index][::2], 0)
+	return time_gluc
+
+def generate_boluses(solution, start_time):
+	""" Generates events for both bolus entries and wizard entries.
+		Returns carb, time and glucose values for each event
+	"""
+	all_carbs = solution[:, 0]
+	glucose = solution[:,1]
+	time = solution[:, 2]
+	ts = make_timesteps(start_time, solution[:,2])
+	positives = []
+	for row in zip(all_carbs, ts, glucose):
+		if row[0] > 10:
+			positives.append(row)
+	np_pos = np.array(clean_up_boluses(positives))
+	cleaned = remove_night_boluses(np_pos)
+	for row in cleaned: #find carb values that are too high and reduce them
+		carb_val = row[0]
+		if carb_val > 120:
+			carb_val= carb_val / 2	
+		elif carb_val > 30:
+			carb_val = carb_val * 0.75
+	bolus, wizard = bolus_or_wizard(cleaned)
+	np_bolus, np_wizard = np.array(bolus), np.array(wizard)
+	b_carbs, b_ts  = np_bolus[:, 0], np_bolus[:, 1]
+	w_card, w_ts, w_gluc = np_wizard[:, 0], np_wizard[:, 1], np_wizard[:,2]
+	return b_carbs, b_ts, w_card, w_ts, w_gluc
+
+def clean_up_boluses(carb_time, filter_rate=7):
+	""" Cleans up clusters of carb events based on meal freqeuncy
+		carb_time -- a numpy array of carbohydrates at different points in time
+		filter_rate -- how many 5 minute segments should be filtered within the range
+					   of a single event.
+					   Example: 7 * 5 --> remove the next 35 minutes of data from carb_time
+	"""
+	return carb_time[::filter_rate]
+
+def remove_night_boluses(carb_time_gluc):
+	"""Removes night boluses excpet for events with high glucose""" 
+	keep = []
+	for row in carb_time_gluc:
+		gluc_val = row[2]
+		hour = time.gmtime(row[1]).tm_hour
+		if (hour > 6 and hour < 23):
+			keep.append(row)
+		elif int(gluc_val) not in range(0,250):
+			keep.append(row)
+	np_keep = np.array(keep)
+	return np_keep
+
+def bolus_or_wizard(solution):
+	"""Randomely decide when to generte wizards events that are linked with boluses 
+	   and when to have plain boluses.
+	   About 2 out of 6 events will be plain boluses
+	"""
+	bolus_events, wizard_events = [], []
+	for row in solution:
+		bolus_or_wizard = random.randint(0, 5)
+		if bolus_or_wizard == 2 or bolus_or_wizard == 4:
+			bolus_events.append(row)
+		else:
+			wizard_events.append(row)
+	return bolus_events, wizard_events
+
+def convert_to_mmol(iterable):
+	conversion_factor = 18.01559
+	return [reading / conversion_factor for reading in iterable]
+
+def round_to(n, precision=0.005):
```

Comment:
Correction is basically to determine if the number is above or below the boundary number -- as in should it be rounded up or rounded down. In the case of rounding it to 0.05 it's dividing it by 20, and int is basically like floor division so it always rounds it down. Then after applying the correction it multiplies it by 20 to get the number back just rounded.